File: snake.rb
# snake.rb
require 'io/console'
require 'timeout'

class SnakeGame
  attr_reader :rows, :cols

  def initialize(rows=21, cols=21)
    @rows, @cols = rows, cols
    @maze = Array.new(rows) { Array.new(cols, '#') }
    @snake = []
    @dir = [0, 1]
    @food = nil
    @score = 0
    @speed = 0.15
    @game_over = false
    @running = true
    generate_maze
    spawn_snake
    spawn_food
    Thread.new { input_loop }
    game_loop
  end

  def generate_maze
    stack = [[1, 1]]
    @maze[1][1] = ' '
    dirs = [[-2,0], [2,0], [0,-2], [0,2]]
    until stack.empty?
      r, c = stack.last
      neigh = []
      dirs.each_with_index do |d, idx|
        nr, nc = r + d[0], c + d[1]
        if nr > 0 && nr < @rows-1 && nc > 0 && nc < @cols-1 && @maze[nr][nc] == '#'
          neigh << idx
        end
      end
      if neigh.any?
        idx = neigh.sample
        dr, dc = dirs[idx]
        nr, nc = r + dr, c + dc
        @maze[r + dr/2][c + dc/2] = ' '
        @maze[nr][nc] = ' '
        stack.push([nr, nc])
      else
        stack.pop
      end
    end
    (0...@rows).each { |i| @maze[i][0] = '#'; @maze[i][@cols-1] = '#' }
    (0...@cols).each { |j| @maze[0][j] = '#'; @maze[@rows-1][j] = '#' }
  end

  def spawn_snake
    cr, cc = @rows/2, @cols/2
    if @maze[cr][cc] != ' '
      (1...@rows-1).each do |r|
        (1...@cols-1).each do |c|
          if @maze[r][c] == ' '
            cr, cc = r, c
            break
          end
        end
        break if @maze[cr][cc] == ' '
      end
    end
    @snake = [[cr, cc]]
    @dir = [0, 1]
  end

  def spawn_food
    empty = []
    (1...@rows-1).each do |r|
      (1...@cols-1).each do |c|
        if @maze[r][c] == ' ' && !@snake.include?([r, c])
          empty << [r, c]
        end
      end
    end
    if empty.any?
      @food = empty.sample
    else
      @game_over = true
    end
  end

  def update
    return if @game_over
    head = @snake.first
    nr, nc = head[0] + @dir[0], head[1] + @dir[1]
    if nr <= 0 || nr >= @rows-1 || nc <= 0 || nc >= @cols-1 || @maze[nr][nc] == '#'
      @game_over = true
      return
    end
    if @snake.include?([nr, nc])
      @game_over = true
      return
    end
    @snake.unshift([nr, nc])
    if [nr, nc] == @food
      @score += 1
      @speed = [0.05, @speed - 0.005].max
      spawn_food
      return if @game_over
    else
      @snake.pop
    end
  end

  def draw
    system('clear') || system('cls')
    puts "Score: #{@score}  Speed: #{ (1/@speed).to_i }"
    display = @maze.map(&:dup)
    @snake.each { |r,c| display[r][c] = 'O' }
    display[@food[0]][@food[1]] = '*' if @food
    display[@snake.first[0]][@snake.first[1]] = 'X'
    display.each { |row| puts row.join }
    puts "GAME OVER! Press R to restart, Q to quit." if @game_over
  end

  def input_loop
    while @running
      char = STDIN.getch
      case char
      when 'q', 'Q' then @running = false; exit
      when 'r', 'R'
        if @game_over
          # restart: reinitialize
          initialize(@rows, @cols)
          return
        end
      when 'w' then @dir = [-1, 0] if @dir != [1, 0]
      when 's' then @dir = [1, 0] if @dir != [-1, 0]
      when 'a' then @dir = [0, -1] if @dir != [0, 1]
      when 'd' then @dir = [0, 1] if @dir != [0, -1]
      when "\e" # arrow keys
        c2 = STDIN.read_nonblock(2) rescue nil
        if c2 == '[A' then @dir = [-1, 0] if @dir != [1, 0]
        elsif c2 == '[B' then @dir = [1, 0] if @dir != [-1, 0]
        elsif c2 == '[C' then @dir = [0, 1] if @dir != [0, -1]
        elsif c2 == '[D' then @dir = [0, -1] if @dir != [0, 1]
        end
      end
    end
  end

  def game_loop
    while @running
      update
      draw
      sleep @speed
    end
  end
end

SnakeGame.new

File: Snake.java
// Snake.java
import java.io.*;
import java.util.*;
import java.util.concurrent.*;

public class Snake {
    static final int ROWS = 21, COLS = 21;
    char[][] maze;
    List<int[]> snake = new ArrayList<>();
    int[] dir = {0, 1};
    int[] food;
    int score = 0;
    int speed = 150; // ms
    boolean gameOver = false;
    boolean running = true;
    Random rand = new Random();

    public Snake() {
        generateMaze();
        spawnSnake();
        spawnFood();
        new Thread(this::handleInput).start();
        gameLoop();
    }

    void generateMaze() {
        maze = new char[ROWS][COLS];
        for (int i = 0; i < ROWS; i++) Arrays.fill(maze[i], '#');
        Stack<int[]> stack = new Stack<>();
        stack.push(new int[]{1, 1});
        maze[1][1] = ' ';
        int[][] dirs = {{-2,0},{2,0},{0,-2},{0,2}};
        while (!stack.isEmpty()) {
            int[] cur = stack.peek();
            List<Integer> neigh = new ArrayList<>();
            for (int i=0; i<4; i++) {
                int nr = cur[0]+dirs[i][0], nc = cur[1]+dirs[i][1];
                if (nr>0 && nr<ROWS-1 && nc>0 && nc<COLS-1 && maze[nr][nc]=='#')
                    neigh.add(i);
            }
            if (!neigh.isEmpty()) {
                int idx = neigh.get(rand.nextInt(neigh.size()));
                int dr = dirs[idx][0], dc = dirs[idx][1];
                int nr = cur[0]+dr, nc = cur[1]+dc;
                maze[cur[0]+dr/2][cur[1]+dc/2] = ' ';
                maze[nr][nc] = ' ';
                stack.push(new int[]{nr, nc});
            } else {
                stack.pop();
            }
        }
        for (int i=0; i<ROWS; i++) { maze[i][0]='#'; maze[i][COLS-1]='#'; }
        for (int j=0; j<COLS; j++) { maze[0][j]='#'; maze[ROWS-1][j]='#'; }
    }

    void spawnSnake() {
        int cr=ROWS/2, cc=COLS/2;
        if (maze[cr][cc] != ' ') {
            outer: for (int r=1; r<ROWS-1; r++)
                for (int c=1; c<COLS-1; c++)
                    if (maze[r][c]==' ') { cr=r; cc=c; break outer; }
        }
        snake.clear();
        snake.add(new int[]{cr, cc});
        dir = new int[]{0, 1};
    }

    void spawnFood() {
        List<int[]> empty = new ArrayList<>();
        for (int r=1; r<ROWS-1; r++)
            for (int c=1; c<COLS-1; c++) {
                if (maze[r][c]==' ') {
                    boolean occ = false;
                    for (int[] p : snake) if (p[0]==r && p[1]==c) { occ=true; break; }
                    if (!occ) empty.add(new int[]{r, c});
                }
            }
        if (!empty.isEmpty()) food = empty.get(rand.nextInt(empty.size()));
        else gameOver = true;
    }

    void update() {
        if (gameOver) return;
        int[] head = snake.get(0);
        int nr = head[0] + dir[0], nc = head[1] + dir[1];
        if (nr<=0 || nr>=ROWS-1 || nc<=0 || nc>=COLS-1 || maze[nr][nc]=='#') {
            gameOver = true; return;
        }
        for (int[] p : snake) if (p[0]==nr && p[1]==nc) { gameOver=true; return; }
        snake.add(0, new int[]{nr, nc});
        if (nr==food[0] && nc==food[1]) {
            score++;
            speed = (int)Math.max(50, speed*0.95);
            spawnFood();
            if (gameOver) return;
        } else {
            snake.remove(snake.size()-1);
        }
    }

    void draw() {
        System.out.print("\033[H\033[2J");
        System.out.printf("Score: %d  Speed: %d\n", score, 1000/speed);
        char[][] disp = new char[ROWS][COLS];
        for (int i=0; i<ROWS; i++) System.arraycopy(maze[i], 0, disp[i], 0, COLS);
        for (int[] p : snake) disp[p[0]][p[1]] = 'O';
        if (!gameOver) disp[food[0]][food[1]] = '*';
        disp[snake.get(0)[0]][snake.get(0)[1]] = 'X';
        for (char[] row : disp) System.out.println(new String(row));
        if (gameOver) System.out.println("GAME OVER! Press R to restart, Q to quit.");
    }

    void handleInput() {
        try {
            while (running) {
                int ch = System.in.read();
                if (ch == -1) continue;
                char c = (char) ch;
                if (c == 'q' || c == 'Q') { running = false; System.exit(0); }
                if (c == 'r' || c == 'R') {
                    if (gameOver) { restart(); return; }
                }
                if (!gameOver) {
                    if (c == 'w' && dir[0] != 1) dir = new int[]{-1, 0};
                    else if (c == 's' && dir[0] != -1) dir = new int[]{1, 0};
                    else if (c == 'a' && dir[1] != 1) dir = new int[]{0, -1};
                    else if (c == 'd' && dir[1] != -1) dir = new int[]{0, 1};
                }
            }
        } catch (IOException e) {}
    }

    void restart() {
        // Reinitialize the game state
        generateMaze();
        spawnSnake();
        spawnFood();
        score = 0;
        speed = 150;
        gameOver = false;
    }

    void gameLoop() {
        while (running) {
            update();
            draw();
            try { Thread.sleep(speed); } catch (InterruptedException e) {}
        }
    }

    public static void main(String[] args) {
        new Snake();
    }
}
